misc/launchd/org.nixos.nix-daemon.plist.in

```diff
@@ -12,14 +12,16 @@
     <key>Label</key>
     <string>org.nixos.nix-daemon</string>
     <key>KeepAlive</key>
-    <true/>
-    <key>RunAtLoad</key>
-    <true/>
+    <dict>
+      <key>OtherJobEnabled</key>
```

Comment:
clever, this effectively prevents the deamon from starting without nix store existing (replacing the need for wait4path?)

Reply:
That's the idea. I think there may be a better way, as `man launchd.plist` mentions that OtherJobsEnabled is to be avoided. There is a RunAtLoad replacement called WatchPaths which I would like to explore just a little more before this should be considered done.

Reply:
Yeah I don't think PathState or WatchPaths really work here. It looks like launchd wants to eagerly check the path to the daemon exists when they are used (even without RunAtLoad). Could be the way I am using it, of course.

Reply:
Does this still work if you don’t have a /nix/store volume? Say you’re still on macOS X?

Reply:
Hm. Good question. I think the installer is making the volume and the darwin-store service based on this variable https://github.com/NixOS/nix/blob/fb662e0acfd8396fdbd4a46a286fec30f261eb92/scripts/install-darwin-multi-user.sh#L8 and this check https://github.com/NixOS/nix/blob/fb662e0acfd8396fdbd4a46a286fec30f261eb92/scripts/install-darwin-multi-user.sh#L39

Reply:
I don't think it'll set up the volume (or the service) if the root is writable (pre-catalina).

I've wondered (but not looked into...) whether it's feasible to bump all new installs up to the volume even where it isn't necessary. I guess 3 small cases for it are:
1. If case-sensitive APFS worked well, I guess it'd let us transition a little faster. I did make the FS overridable so that anyone feeling adventurous could play with it, but I'm not sure if anyone has.
2. If it doesn't cause its own unique issues, reducing the variation across the install base might pay small support/troubleshooting dividends over time.
3. I assume most people still running versions that old are doing it on purpose, but a more normal modern install might protect them from a break if they decide to upgrade macOS

Reply:
I remember during the original macOS encrypted nix install discussion there was the goal of getting all new macOS installs into the same configuration, to try and make everyone behave the same and ensure the same instructions work everywhere.

Reply:
In any case, I'm skeptical that `OtherJobEnabled` is sufficient here. The docs say this keeps this job alive as long as the other job is loaded. It does _not_ say it allows the other job to run to completion first. It's entirely reasonable for launchd to load both jobs, spawn the nix store mounter, and then attempt to spawn `nix-daemon` before the nix store has finished mounting.

The docs of `PathState` make it sound like that actually does pretty much what we want, keeping `nix-daemon` alive as long as the given path exists (which would presumably be the path to `nix-daemon`). I'm curious to know what the issue is with it. The docs do encourage using other alternatives as "Filesystem monitoring mechanisms are inherently race-prone and lossy", but we don't care about responding to all filesystem events. I would certainly hope that "a volume is mounted that has a given path" is a reliable trigger for filesystem monitoring.

Reply:
In my testing (flailing, really), whenever I tried just PathState I would get logs of this sort and the daemon would have not started with code 78:

```
2021-12-01 07:57:56.953243 (system/org.nixos.nix-daemon [97]) <Notice>: xpcproxy spawned with pid 97                                                                                                               
2021-12-01 07:57:56.953259 (system/org.nixos.nix-daemon [97]) <Notice>: internal event: SPAWNED, code = 0                                                                                                          
2021-12-01 07:57:56.953262 (system/org.nixos.nix-daemon [97]) <Notice>: service state: xpcproxy                                                                                                                    
2021-12-01 07:57:56.963433 (system/org.nixos.nix-daemon [97]) <Notice>: internal event: SOURCE_ATTACH, code = 0                                                                                                    
2021-12-01 07:57:57.533010 (system/org.nixos.nix-daemon [97]) <Warning>: Could not find and/or execute program specified by service: 2: No such file or directory: /nix/var/nix/profiles/default/bin/nix-daemon    
2021-12-01 07:57:57.533023 (system/org.nixos.nix-daemon [97]) <Error>: Service could not initialize: posix_spawn(/nix/var/nix/profiles/default/bin/nix-daemon) not accessible error: 0x6f: Invalid or missing Progr
am/ProgramArguments                                                                                                                                                                                                
2021-12-01 07:57:57.533027 (system/org.nixos.nix-daemon [97]) <Error>: initialization failure: 21A559: xpcproxy + 23780 [840][F29643C9-8E6C-3632-93A1-5214FFD1DC57]: 0x6f                                          
2021-12-01 07:57:57.533029 (system/org.nixos.nix-daemon [97]) <Notice>: Service setup event to handle failure and will not launch until it fires.                                                                  
2021-12-01 07:57:57.533031 (system/org.nixos.nix-daemon [97]) <Error>: Missing executable detected. Job: 'org.nixos.nix-daemon' Executable: '/nix/var/nix/profiles/default/bin/nix-daemon'                         
2021-12-01 07:57:57.533033 (system/org.nixos.nix-daemon [97]) <Notice>: internal event: INIT, code = 111                                                                                                           
2021-12-01 07:57:57.534043 (system/org.nixos.nix-daemon [97]) <Notice>: trampoline exited with code: 78                                                                                                            
2021-12-01 07:57:57.534046 (system/org.nixos.nix-daemon [97]) <Notice>: service exited: dirty = 0, supported pressured-exit = 0                                                                                    
2021-12-01 07:57:57.534049 (system/org.nixos.nix-daemon [97]) <Notice>: exited due to exit(78)                                                                                                                     
2021-12-01 07:57:57.534051 (system/org.nixos.nix-daemon [97]) <Notice>: already handled failed init, ignoring                                                                                                      
2021-12-01 07:57:57.534054 (system/org.nixos.nix-daemon [97]) <Notice>: service state: exited                                                                                                                      
2021-12-01 07:57:57.534057 (system/org.nixos.nix-daemon [97]) <Notice>: internal event: EXITED, code = 0 
```

Reply:
Are there any logs related to when the `/nix` volume mounted?

Reply:
Interesting. This suggests that the mount command run by `org.nixos.darwin-store` perhaps exited prior to the actual filesystem being available. That seems rather surprising. I suppose this could perhaps be hacked by having this job follow it up with a `wait4path` on the store so the job doesn't exit until the store is actually reachable.

Reply:
@lilyball Are you talking about this line?
```
2021-12-02 11:17:53.747178 (system/org.nixos.darwin-store [247]) <Notice>: Successfully spawned diskutil[247] because xpc event
```

Reply:
> This suggests that the mount command run by org.nixos.darwin-store perhaps exited prior to the actual filesystem being available. That seems rather surprising.

Is there anything specific to suggest that is more likely than the PathState trigger having some latency? Do we know if the mechanism is just polling? The manpage does suggest it's both race-prone and lossy. I'm not certain what lossy means here, but my first guess would be that it might miss filesystem conditions that don't persist for longer than some polling interval?

Reply:
> Is there anything specific to suggest that is more likely than the PathState trigger having some latency?

Perhaps I misunderstood, but it looks like in your second log that it runs `diskutil`, that exits, it unloads the job, and then spawns nix-daemon. And in the first log the nix-daemon spawn fails because `/nix/var/nix/profiles/default/bin/nix-daemon` isn't accessible yet. Latency wouldn't cause the path to become inaccessible. So my impression was that `diskutil` must have exited prior to the filesystem actually being accessible and therefore launchd tried to launch nix-daemon too soon. Though that doesn't answer the question of why the nix-daemon job would have launched at all given that PathState means it shouldn't launch until the path is accessible.

Though looking at the second log now, there's a delay of several seconds in between diskutil exiting and nix-daemon being launched. So it's clearly waiting for something. PathState having latency would explain that delay (which could also just be launchd prioritizing other work prior to responding to PathState), but doesn't explain why the first log failed.

Do you have a log of a _failure_ that includes the `org.nixos.darwin-store` lines?

Reply:
@lilyball sure. I need a little bit to go setup a vm for myself

Reply:
On commit `5d959b33c5a75f3053280a8a34c711f964437766` without filevault enabled.
```
sh <(curl -L https://jsoo1-nix-install-tests.cachix.org/serve/v89k06b82dkhalcdkdhnfbmrfr6fp1w9/install) --tarball-url-prefix https://jsoo1-nix-install-tests.cachix.org/serve
```
<details>
<summary>The first install goes well:</summary>

```ShellSession
% grep nixos /var/log/com.apple.xpc.launchd/launchd.log
2021-12-03 14:43:25.520241 (system/org.nixos.darwin-store) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 14:43:25.520245 (system/org.nixos.darwin-store) <Notice>: service state: spawn scheduled
2021-12-03 14:43:25.520246 (system/org.nixos.darwin-store) <Notice>: service state: spawning
2021-12-03 14:43:25.520307 (system/org.nixos.darwin-store) <Notice>: launching: speculative
2021-12-03 14:43:25.521383 (system/org.nixos.darwin-store [740]) <Notice>: xpcproxy spawned with pid 740
2021-12-03 14:43:25.521399 (system/org.nixos.darwin-store [740]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 14:43:25.521401 (system/org.nixos.darwin-store [740]) <Notice>: service state: xpcproxy
2021-12-03 14:43:25.521468 (system) <Notice>: Bootstrap by launchctl[739] for /Library/LaunchDaemons/org.nixos.darwin-store.plist succeeded (0: )
2021-12-03 14:43:25.521506 (system/org.nixos.darwin-store [740]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 14:43:25.526997 (system/org.nixos.darwin-store [740]) <Notice>: service state: running
2021-12-03 14:43:25.527000 (system/org.nixos.darwin-store [740]) <Notice>: internal event: INIT, code = 0
2021-12-03 14:43:25.527005 (system/org.nixos.darwin-store [740]) <Notice>: Successfully spawned diskutil[740] because speculative
2021-12-03 14:43:25.591782 (system/org.nixos.darwin-store [740]) <Notice>: signaled service: Terminated: 15
2021-12-03 14:43:25.591792 (system/org.nixos.darwin-store [740]) <Notice>: service state: SIGTERMed
2021-12-03 14:43:25.591794 (system/org.nixos.darwin-store [740]) <Notice>: scheduling cleanup in 5 sec after sending Terminated: 15
2021-12-03 14:43:25.592549 (system/org.nixos.darwin-store [740]) <Notice>: service exited: dirty = 0, supported pressured-exit = 0
2021-12-03 14:43:25.592551 (system/org.nixos.darwin-store [740]) <Notice>: exited due to SIGTERM | sent by launchd[1]
2021-12-03 14:43:25.592553 (system/org.nixos.darwin-store [740]) <Notice>: service state: exited
2021-12-03 14:43:25.592556 (system/org.nixos.darwin-store [740]) <Notice>: internal event: EXITED, code = 0
2021-12-03 14:43:25.592558 (system) <Notice>: service inactive: org.nixos.darwin-store
2021-12-03 14:43:25.592560 (system/org.nixos.darwin-store [740]) <Notice>: service state: not running
2021-12-03 14:43:25.592582 (system/org.nixos.darwin-store) <Notice>: Service only ran for 0 seconds. Pushing respawn out by 10 seconds.
2021-12-03 14:43:25.592586 (system/org.nixos.darwin-store) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 14:43:25.592587 (system/org.nixos.darwin-store) <Notice>: service state: spawn scheduled
2021-12-03 14:43:25.592589 (system/org.nixos.darwin-store) <Notice>: service throttled by 10 seconds
2021-12-03 14:43:35.598100 (system/org.nixos.darwin-store) <Notice>: service state: spawning
2021-12-03 14:43:35.598154 (system/org.nixos.darwin-store) <Notice>: launching: non-ipc demand
2021-12-03 14:43:35.598730 (system/org.nixos.darwin-store [746]) <Notice>: xpcproxy spawned with pid 746
2021-12-03 14:43:35.598746 (system/org.nixos.darwin-store [746]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 14:43:35.598748 (system/org.nixos.darwin-store [746]) <Notice>: service state: xpcproxy
2021-12-03 14:43:35.598750 (system/org.nixos.darwin-store [746]) <Notice>: deferred event: domain spawn response: 0
2021-12-03 14:43:35.598754 (system/org.nixos.darwin-store [746]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 14:43:35.601500 (system/org.nixos.darwin-store [746]) <Notice>: service state: running
2021-12-03 14:43:35.601508 (system/org.nixos.darwin-store [746]) <Notice>: internal event: INIT, code = 0
2021-12-03 14:43:35.601513 (system/org.nixos.darwin-store [746]) <Notice>: Successfully spawned diskutil[746] because non-ipc demand
2021-12-03 14:43:35.647161 (system/org.nixos.darwin-store [746]) <Notice>: job state = running
2021-12-03 14:43:35.873600 (system/org.nixos.darwin-store [746]) <Notice>: service exited: dirty = 0, supported pressured-exit = 0
2021-12-03 14:43:35.873609 (system/org.nixos.darwin-store [746]) <Notice>: exited due to exit(0)
2021-12-03 14:43:35.873612 (system/org.nixos.darwin-store [746]) <Notice>: service state: exited
2021-12-03 14:43:35.873614 (system/org.nixos.darwin-store [746]) <Notice>: internal event: EXITED, code = 0
2021-12-03 14:43:35.873616 (system/org.nixos.darwin-store [746]) <Notice>: job state = exited
2021-12-03 14:43:35.873630 (system) <Notice>: service inactive: org.nixos.darwin-store
2021-12-03 14:43:35.873632 (system/org.nixos.darwin-store [746]) <Notice>: service state: not running
2021-12-03 14:43:35.874726 (system/org.nixos.darwin-store) <Notice>: job is not monitored, can't poll
2021-12-03 14:44:53.708894 (system/org.nixos.nix-daemon) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 14:44:53.708898 (system/org.nixos.nix-daemon) <Notice>: service state: spawn scheduled
2021-12-03 14:44:53.708900 (system/org.nixos.nix-daemon) <Notice>: service state: spawning
2021-12-03 14:44:53.708965 (system/org.nixos.nix-daemon) <Notice>: launching: speculative
2021-12-03 14:44:53.710107 (system/org.nixos.nix-daemon [2512]) <Notice>: xpcproxy spawned with pid 2512
2021-12-03 14:44:53.710124 (system/org.nixos.nix-daemon [2512]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 14:44:53.710127 (system/org.nixos.nix-daemon [2512]) <Notice>: service state: xpcproxy
2021-12-03 14:44:53.710189 (system) <Notice>: Bootstrap by launchctl[2511] for /Library/LaunchDaemons/org.nixos.nix-daemon.plist succeeded (0: )
2021-12-03 14:44:53.710249 (system/org.nixos.nix-daemon [2512]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 14:44:53.719456 (system/org.nixos.nix-daemon [2512]) <Notice>: service state: running
2021-12-03 14:44:53.719460 (system/org.nixos.nix-daemon [2512]) <Notice>: internal event: INIT, code = 0
2021-12-03 14:44:53.719465 (system/org.nixos.nix-daemon [2512]) <Notice>: Successfully spawned nix-daemon[2512] because speculative
2021-12-03 14:44:53.821718 (system/org.nixos.nix-daemon [2512]) <Notice>: signaled service: Terminated: 15
2021-12-03 14:44:53.821730 (system/org.nixos.nix-daemon [2512]) <Notice>: service state: SIGTERMed
2021-12-03 14:44:53.821732 (system/org.nixos.nix-daemon [2512]) <Notice>: scheduling cleanup in 5 sec after sending Terminated: 15
2021-12-03 14:44:53.821902 (system/org.nixos.nix-daemon [2512]) <Notice>: service exited: dirty = 0, supported pressured-exit = 0
2021-12-03 14:44:53.821904 (system/org.nixos.nix-daemon [2512]) <Notice>: exited due to SIGTERM | sent by launchd[1]
2021-12-03 14:44:53.821906 (system/org.nixos.nix-daemon [2512]) <Notice>: service state: exited
2021-12-03 14:44:53.821909 (system/org.nixos.nix-daemon [2512]) <Notice>: internal event: EXITED, code = 0
2021-12-03 14:44:53.821911 (system) <Notice>: service inactive: org.nixos.nix-daemon
2021-12-03 14:44:53.821921 (system/org.nixos.nix-daemon [2512]) <Notice>: service state: not running
2021-12-03 14:44:53.821940 (system/org.nixos.nix-daemon) <Notice>: Service only ran for 0 seconds. Pushing respawn out by 10 seconds.
2021-12-03 14:44:53.821944 (system/org.nixos.nix-daemon) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 14:44:53.821946 (system/org.nixos.nix-daemon) <Notice>: service state: spawn scheduled
2021-12-03 14:44:53.821948 (system/org.nixos.nix-daemon) <Notice>: service throttled by 10 seconds
2021-12-03 14:44:53.821964 (system/org.nixos.nix-daemon) <Notice>: launch already in progress
2021-12-03 14:45:03.827328 (system/org.nixos.nix-daemon) <Notice>: service state: spawning
2021-12-03 14:45:03.827362 (system/org.nixos.nix-daemon) <Notice>: launching: xpc event
2021-12-03 14:45:03.827998 (system/org.nixos.nix-daemon [2518]) <Notice>: xpcproxy spawned with pid 2518
2021-12-03 14:45:03.828014 (system/org.nixos.nix-daemon [2518]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 14:45:03.828016 (system/org.nixos.nix-daemon [2518]) <Notice>: service state: xpcproxy
2021-12-03 14:45:03.828018 (system/org.nixos.nix-daemon [2518]) <Notice>: deferred event: domain spawn response: 0
2021-12-03 14:45:03.828022 (system/org.nixos.nix-daemon [2518]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 14:45:03.832118 (system/org.nixos.nix-daemon [2518]) <Notice>: service state: running
2021-12-03 14:45:03.832126 (system/org.nixos.nix-daemon [2518]) <Notice>: internal event: INIT, code = 0
2021-12-03 14:45:03.832130 (system/org.nixos.nix-daemon [2518]) <Notice>: Successfully spawned nix-daemon[2518] because xpc event
```
After a reboot, though:
```
% grep nixos /var/log/com.apple.xpc.launchd/launchd.log
2021-12-03 14:56:43.145942 (system) <Notice>: pending spawn, domain in on-demand-only mode: org.nixos.darwin-store
2021-12-03 14:56:43.147175 (system) <Notice>: pending spawn, domain in on-demand-only mode: org.nixos.nix-daemon
2021-12-03 14:56:43.167185 (system/org.nixos.darwin-store) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 14:56:43.167188 (system/org.nixos.darwin-store) <Notice>: service state: spawn scheduled
2021-12-03 14:56:43.167190 (system/org.nixos.darwin-store) <Notice>: service state: spawning
2021-12-03 14:56:43.167230 (system/org.nixos.darwin-store) <Notice>: launching: speculative
2021-12-03 14:56:43.168707 (system/org.nixos.darwin-store [70]) <Notice>: xpcproxy spawned with pid 70
2021-12-03 14:56:43.168719 (system/org.nixos.darwin-store [70]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 14:56:43.168721 (system/org.nixos.darwin-store [70]) <Notice>: service state: xpcproxy
2021-12-03 14:56:43.200394 (system/org.nixos.nix-daemon) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 14:56:43.200397 (system/org.nixos.nix-daemon) <Notice>: service state: spawn scheduled
2021-12-03 14:56:43.200399 (system/org.nixos.nix-daemon) <Notice>: service state: spawning
2021-12-03 14:56:43.200421 (system/org.nixos.nix-daemon) <Notice>: launching: speculative
2021-12-03 14:56:43.201162 (system/org.nixos.nix-daemon [99]) <Notice>: xpcproxy spawned with pid 99
2021-12-03 14:56:43.201173 (system/org.nixos.nix-daemon [99]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 14:56:43.201175 (system/org.nixos.nix-daemon [99]) <Notice>: service state: xpcproxy
2021-12-03 14:56:43.233166 (system/org.nixos.darwin-store [70]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 14:56:43.233228 (system/org.nixos.nix-daemon [99]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 14:56:43.495702 (system/org.nixos.nix-daemon [99]) <Warning>: Could not find and/or execute program specified by service: 2: No such file or directory: /nix/var/nix/profiles/default/bin/nix-daemon
2021-12-03 14:56:43.495706 (system/org.nixos.nix-daemon [99]) <Error>: Service could not initialize: posix_spawn(/nix/var/nix/profiles/default/bin/nix-daemon) not accessible error: 0x6f: Invalid or missing Program/ProgramArguments
2021-12-03 14:56:43.495709 (system/org.nixos.nix-daemon [99]) <Error>: initialization failure: 21A559: xpcproxy + 23780 [840][F29643C9-8E6C-3632-93A1-5214FFD1DC57]: 0x6f
2021-12-03 14:56:43.495711 (system/org.nixos.nix-daemon [99]) <Notice>: Service setup event to handle failure and will not launch until it fires.
2021-12-03 14:56:43.495713 (system/org.nixos.nix-daemon [99]) <Error>: Missing executable detected. Job: 'org.nixos.nix-daemon' Executable: '/nix/var/nix/profiles/default/bin/nix-daemon'
2021-12-03 14:56:43.495715 (system/org.nixos.nix-daemon [99]) <Notice>: internal event: INIT, code = 111
2021-12-03 14:56:43.502612 (system/org.nixos.nix-daemon [99]) <Notice>: trampoline exited with code: 78
2021-12-03 14:56:43.502617 (system/org.nixos.nix-daemon [99]) <Notice>: service exited: dirty = 0, supported pressured-exit = 0
2021-12-03 14:56:43.502619 (system/org.nixos.nix-daemon [99]) <Notice>: exited due to exit(78)
2021-12-03 14:56:43.502621 (system/org.nixos.nix-daemon [99]) <Notice>: already handled failed init, ignoring
2021-12-03 14:56:43.502623 (system/org.nixos.nix-daemon [99]) <Notice>: service state: exited
2021-12-03 14:56:43.502625 (system/org.nixos.nix-daemon [99]) <Notice>: internal event: EXITED, code = 0
2021-12-03 14:56:43.502627 (system) <Notice>: service inactive: org.nixos.nix-daemon
2021-12-03 14:56:43.502641 (system/org.nixos.nix-daemon [99]) <Notice>: service state: not running
2021-12-03 14:56:43.502643 (system/org.nixos.nix-daemon) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 14:56:43.502647 (system/org.nixos.nix-daemon) <Notice>: service state: spawn scheduled
2021-12-03 14:56:43.972199 (system/org.nixos.darwin-store [70]) <Notice>: service state: running
2021-12-03 14:56:43.972203 (system/org.nixos.darwin-store [70]) <Notice>: internal event: INIT, code = 0
2021-12-03 14:56:43.972206 (system/org.nixos.darwin-store [70]) <Notice>: Successfully spawned diskutil[70] because speculative
2021-12-03 14:56:52.950907 (system/org.nixos.darwin-store [70]) <Notice>: job state = running
2021-12-03 14:56:53.302459 (system/org.nixos.darwin-store [70]) <Notice>: service exited: dirty = 0, supported pressured-exit = 0
2021-12-03 14:56:53.302461 (system/org.nixos.darwin-store [70]) <Notice>: exited due to exit(0)
2021-12-03 14:56:53.302463 (system/org.nixos.darwin-store [70]) <Notice>: service state: exited
2021-12-03 14:56:53.302465 (system/org.nixos.darwin-store [70]) <Notice>: internal event: EXITED, code = 0
2021-12-03 14:56:53.302467 (system/org.nixos.darwin-store [70]) <Notice>: job state = exited
2021-12-03 14:56:53.302481 (system) <Notice>: service inactive: org.nixos.darwin-store
2021-12-03 14:56:53.302483 (system/org.nixos.darwin-store [70]) <Notice>: service state: not running
```
</details>
Note that that commit has `RunAtLoad = true`.  If I `sudo launchctl bootout system /Library/LaunchDaemons/org.nixos.nix-daemon.plist` then change `RunAtLoad = false` and `sudo launchctl bootstram system /Library/LaunchDaemons/org.nixos.nix-daemon.plist` and reboot:

<details>
<summary>No RunAtLoad</summary>

```
2021-12-03 15:07:52.156447 (system) <Notice>: pending spawn, domain in on-demand-only mode: org.nixos.darwin-store
2021-12-03 15:07:52.177604 (system/org.nixos.darwin-store) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 15:07:52.177606 (system/org.nixos.darwin-store) <Notice>: service state: spawn scheduled
2021-12-03 15:07:52.177608 (system/org.nixos.darwin-store) <Notice>: service state: spawning
2021-12-03 15:07:52.177633 (system/org.nixos.darwin-store) <Notice>: launching: speculative
2021-12-03 15:07:52.178811 (system/org.nixos.darwin-store [70]) <Notice>: xpcproxy spawned with pid 70
2021-12-03 15:07:52.178823 (system/org.nixos.darwin-store [70]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 15:07:52.178825 (system/org.nixos.darwin-store [70]) <Notice>: service state: xpcproxy
2021-12-03 15:07:52.241388 (system/org.nixos.darwin-store [70]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 15:07:52.605618 (system/org.nixos.darwin-store [70]) <Notice>: service state: running
2021-12-03 15:07:52.605639 (system/org.nixos.darwin-store [70]) <Notice>: internal event: INIT, code = 0
2021-12-03 15:07:52.605643 (system/org.nixos.darwin-store [70]) <Notice>: Successfully spawned diskutil[70] because speculative
2021-12-03 15:08:01.843502 (system/org.nixos.darwin-store [70]) <Notice>: job state = running
2021-12-03 15:08:01.988518 (system/org.nixos.nix-daemon) <Notice>: internal event: WILL_SPAWN, code = 0
2021-12-03 15:08:01.988533 (system/org.nixos.nix-daemon) <Notice>: service state: spawn scheduled
2021-12-03 15:08:01.988535 (system/org.nixos.nix-daemon) <Notice>: service state: spawning
2021-12-03 15:08:01.988778 (system/org.nixos.nix-daemon) <Notice>: launching: xpc event
2021-12-03 15:08:01.990454 (system/org.nixos.nix-daemon [233]) <Notice>: xpcproxy spawned with pid 233
2021-12-03 15:08:01.990467 (system/org.nixos.nix-daemon [233]) <Notice>: internal event: SPAWNED, code = 0
2021-12-03 15:08:01.990470 (system/org.nixos.nix-daemon [233]) <Notice>: service state: xpcproxy
2021-12-03 15:08:01.990514 (system/org.nixos.nix-daemon [233]) <Notice>: internal event: SOURCE_ATTACH, code = 0
2021-12-03 15:08:02.010866 (system/org.nixos.nix-daemon [233]) <Notice>: service state: running
2021-12-03 15:08:02.010892 (system/org.nixos.nix-daemon [233]) <Notice>: internal event: INIT, code = 0
2021-12-03 15:08:02.010896 (system/org.nixos.nix-daemon [233]) <Notice>: Successfully spawned nix-daemon[233] because xpc event
2021-12-03 15:08:02.057506 (system/org.nixos.darwin-store [70]) <Notice>: service exited: dirty = 0, supported pressured-exit = 0
2021-12-03 15:08:02.057508 (system/org.nixos.darwin-store [70]) <Notice>: exited due to exit(0)
2021-12-03 15:08:02.057510 (system/org.nixos.darwin-store [70]) <Notice>: service state: exited
2021-12-03 15:08:02.057512 (system/org.nixos.darwin-store [70]) <Notice>: internal event: EXITED, code = 0
2021-12-03 15:08:02.057514 (system/org.nixos.darwin-store [70]) <Notice>: job state = exited
2021-12-03 15:08:02.057526 (system) <Notice>: service inactive: org.nixos.darwin-store
2021-12-03 15:08:02.057528 (system/org.nixos.darwin-store [70]) <Notice>: service state: not running
```
</details>

I thought I had tested that configuration and experienced a failure, but I may not have. Seems like `PathState` for the nix-daemon may be all that is required.

Reply:
I pushed the change reverting `OtherJobsEnabled`.